Transaction Builder

stefan-mysten opened the pull request:
This is the first iteration on the transaction builder. I expect that it is not 100% correct as there are some areas that I am not sure about, thus I decided to push a PR to get some feedback. 

Might be easier to go over it in a call, so we can also do that.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+
+    /// Takes an unresolved transaction and attempts to resolve it by querying the GraphQL service.
+    pub fn resolve_transaction(
+        &self,
```
This needs to take in a `client` as well, to query the RPC for resolving this transaction.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+            ix,
+            match self {
+                Value::Result(i) => *i,
+                _ => panic!("Cannot nest a non-result value"),
```
This panic probably needs to be handled better?

stefan-mysten replied:
Fixed

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+    fn from(value: Value) -> Input {
+        match value {
+            Value::Input(i) => Input::Pure(bcs::to_bytes(&i.to_be_bytes()).unwrap()),
+            _ => panic!("Cannot convert Value to Input"),
```
This panic needs to be handled better. Probably we need a `TryFrom` here.

stefan-mysten replied:
Indeed, I don't really understand what I was trying to do.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+        (address, pk)
+    }
+
+    /// Generate a private key and its corresponding address, call faucet which returns 5 coin
```
This assumes that a local network with a faucet is used...

stefan-mysten replied:
Fixed this one.

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+    }
+}
+
+impl From<Input> for UnresolvedInputArgument {
```
Similar question as for `Value` and `Argument` on whether these types can be merged -- they are intended to represent the same domain, so it makes sense that you can implement conversions between them, but maybe we can avoid that entirely.

stefan-mysten replied:
I am thinking how we can do it, but what about `Literals`, when we get tx resolution?

amnn replied:
Literals are specific to transaction building, there is not an analogous type in the types crate, but that's fine because we just implement `From<Literal> for Input` (or whatever I suggested we should use instead of `Input` above).

stefan-mysten replied:
I implemented this as well, except the literal part. Tests pass but I am not 100% confident it's correct.

amnn commented on crates/sui-transaction-builder/src/object.rs:
```diff
@@ -0,0 +1,203 @@
+    Shared,
+}
+
+impl Object {
```
More docs here would be good

stefan-mysten replied:
Added more docs.

amnn commented on crates/sui-transaction-builder/src/object.rs:
```diff
@@ -0,0 +1,203 @@
+
+    // Kind
+    pub fn as_owned(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::ImmOrOwned),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
```
(and same for all the rest below)

```suggestion
        Self { kind: Some(Kind::ImmOrOwned), ..self }
```

stefan-mysten replied:
neat!

amnn commented on crates/sui-transaction-builder/src/object.rs:
```diff
@@ -0,0 +1,203 @@
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: Some(false),
```
Should be `None` (and we should do the same for `initial_shared_version`, and we should do something similar in the `as_owned` impl)

amnn commented on crates/sui-transaction-builder/src/object.rs:
```diff
@@ -0,0 +1,203 @@
+            kind: Some(Kind::Receiving),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
```
Should set these to `None` as well.

amnn commented on crates/sui-transaction-builder/src/object.rs:
```diff
@@ -0,0 +1,203 @@
+        Self {
+            id: self.id,
+            kind: Some(Kind::Shared),
+            version: self.version,
+            digest: self.digest,
```
These should be set to `None`.

amnn commented on crates/sui-transaction-builder/src/object.rs:
```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    // Shared value mutability
+    pub fn by_val(self) -> Self {
+        self
+    }
+    pub fn by_ref(&self) -> &Self {
+        self
+    }
+    pub fn by_mut(&mut self) -> &mut Self {
+        self
+    }
```
```suggestion
    // Shared value mutability
    pub fn by_val(self) -> Self {
        Self { mutable: Some(true), ..self }
    }
    pub fn by_ref(&self) -> &Self {
        Self { mutable: Some(false), ..self }
    }
    pub fn by_mut(&mut self) -> &mut Self {
        Self { mutable: Some(true), ..self }
    }
```

stefan-mysten replied:
I don't fully follow these functions. How would these be used?

amnn replied:
They are used to set the mutability of a shared object input, based on whether that input is being used by value, by reference, or by mutable reference.

amnn commented on crates/sui-transaction-builder/src/object.rs:
```diff
@@ -0,0 +1,203 @@
+    }
+}
+
+impl TryInto<ObjectReference> for &Object {
+    type Error = anyhow::Error;
+
+    fn try_into(self) -> Result<ObjectReference, Self::Error> {
+        Ok(ObjectReference::new(
+            self.id,
+            self.version.ok_or_else(|| anyhow!("version not set"))?,
+            self.digest.ok_or_else(|| anyhow!("digest not set"))?,
+        ))
+    }
+}
```
Generally `From` impls are preferable to `Into` impls [because `From` implies `Into` but not the other way around](https://doc.rust-lang.org/rust-by-example/conversion/from_into.html#from-and-into-are-interchangable).

And generally, you should avoid conversion operators that take references -- they suggest that there isn't really a conversion going on.

stefan-mysten replied:
Yeah, don't know why I did not do that. 🙃

stefan-mysten commented:
Thanks so much for the review @amnn 🚀

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+            gas: Vec::new(),
+            gas_budget: 0,
+            gas_price: 0,
+            sender: Address::default(),
```
`Address` has a `Default` derive on it. Not sure why, @bmwill  probably has more insights.

stefan-mysten replied:
It does seems thought that tests pass without `Default`. Not sure how this impacts the usage in the main `sui` monorepo.

stefan-mysten replied:
Switched to `Option`

bmwill replied:
Happy to remove the Default on Address, there isn't a strong reason to have it imo

stefan-mysten replied:
I'll add it to this PR.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+        let cmd = Command::MoveCall(MoveCall {
+            arguments: arguments.into_iter().map(|a| a.into()).collect(),
+            package: function.package.into(),
+            module: Identifier::new(function.module).unwrap(),
```
Agreed!

bmwill replied:
yep we do have Identifier in the SDK :)

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+
+impl Value {
+    /// Turn a Result into a NestedResult.
+    fn _nested(&self, ix: u16) -> Value {
```
It was not being used anywhere because I did not make it public, so clippy would complain.

amnn replied:
But it should also be public, right?

stefan-mysten replied:
Yes, fixed that.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+    }
+}
+
+// this seems to be needed as without it this is not possible
+// let input = tx.input(Serialized(&1u64).serialize());
+impl From<Vec<u8>> for Input {
+    fn from(bytes: Vec<u8>) -> Input {
+        Input::Pure(bytes)
+    }
+}
+
+impl<'a, T: Serialize> Serialized<'a, T> {
+    /// Serializes the given value into bcs bytes.
+    pub fn serialize(&self) -> Vec<u8> {
+        bcs::to_bytes(self.0).unwrap()
+    }
+}
```
Yes, I was quite confused about this one, did not figure out that this needs to be done `for Input`. 🤦 .
This indeed worked, thanks for the suggestion.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+}
+
+// Allow converting Serialized<Address> into Input
+impl<'a> From<Serialized<'a, Address>> for Input {
+    fn from(serialized: Serialized<'a, Address>) -> Input {
+        // Convert Serialized<Address> into Input::Pure variant
+        Input::Pure(serialized.serialize()) // Here we use `Pure(Vec<u8>)`
+    }
+}
```
Don't need this anymore.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+                        }),
+                    }
+                } else {
+                    Err(anyhow!("Object kind not found"))
```
Yep, switching to custom errors.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+    /// Try to finish the transaction, but if not all parts are resolved, it will return an
+    /// [`UnresolvedTransaction`]. This can be used to resolve the transaction later by calling the
+    /// [`resolve_transaction`] method.
+    pub fn try_finish(&self) -> Result<TransactionResolution, Error> {
```
Simplified

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+
+    /// Assuming everything is resolved, convert this transaction into the
+    /// resolved form. Fails if there are unresolved parts.
+    pub fn finish(&self) -> Result<Transaction, Error> {
```
Simplified.

stefan-mysten replied:
However, I had to add a `self.clone()` in another method, so will have to look again at it to make sure it's ok.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+    }
+}
+
+/// Convert a [`Value`] into a transaction command [`Argument`] type.
```
We can. The only thing is we need to add the `nested` function in an `impl` in that crate. Added.

bmwill commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+    // Commands
+    /// Call a Move function with the given arguments.
+    pub fn move_call(&mut self, function: Function, arguments: Vec<Value>) -> Value {
+        let cmd = Command::MoveCall(MoveCall {
```
which is how they are defined in the struct itself

stefan-mysten replied:
Fixed!

stefan-mysten commented on crates/sui-sdk-types/src/types/address.rs:
```diff
@@ -1,4 +1,4 @@
-#[derive(Clone, Copy, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
```
We don't need a Default on an Address, and it's probably best not to have it.

stefan-mysten replied:
Removed

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,917 @@
+        self.expiration = Some(TransactionExpiration::Epoch(epoch));
+    }
+
+    // Commands
```
Added more explanations.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1043 @@
+                let Some(client) = client else {
+                    return Err(Error::NoClient);
+                };
+                let Some(sender) = self.sender else {
```
Hmm, I assume that the transaction resolution API could infer the sender based on the inputs?

amnn replied:
I guess it could, technically? I.e. if there are owned objects then you could infer the sender from them, but this is not possible in all cases (because a transaction may have no owned objects apart from the gas, which could be owned by a sponsor who is different from the sender), and in general this seems more confusing than helpful -- it should be fine to expect that the transaction always has a sender supplied before you try to finish it.

stefan-mysten replied:
Thanks for input here. We keep this check.

stefan-mysten commented on crates/sui-sdk-types/src/types/transaction/mod.rs:
```diff
@@ -626,6 +626,19 @@ pub enum Argument {
+impl Argument {
+    /// Turn a Result into a NestedResult.
+    pub fn nested(&self, ix: u16) -> Argument {
+        Argument::NestedResult(
+            match self {
+                Argument::Result(i) => *i,
+                _ => panic!("Cannot nest a non-result value"),
+            },
+            ix,
```
Switch to let else for nicer code.

stefan-mysten commented on crates/sui-transaction-builder/src/error.rs:
```diff
@@ -0,0 +1,34 @@
+    NoGasPrice,
+    #[error("No commands provided, only inputs")]
+    NoCommandsOnlyInputs,
+    #[error("No inputs provided, only commands")]
+    NoInputsOnlyCommands,
```
We can have commands which take no inputs. This needs to go!

stefan-mysten commented on crates/sui-transaction-builder/src/error.rs:
```diff
@@ -0,0 +1,34 @@
+    NoGasBudget,
+    #[error("No gas price provided.")]
+    NoGasPrice,
+    #[error("No commands provided, only inputs")]
+    NoCommandsOnlyInputs,
```
This can also be removed. A transaction with only inputs is valid.

stefan-mysten commented on crates/sui-transaction-builder/src/error.rs:
```diff
@@ -0,0 +1,34 @@
+    #[error("Initial shared version not found.")]
+    InitialSharedVersionNotFound,
+    #[error("Mutable property not found.")]
+    MutablePropertyNotFound,
```
Rename this error and change the error message; _could not determine mutability of this shared object._

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+    ///   for missing data. A client is required for this operation.
+    ///
+    /// It returns an error if the transaction cannot be resolved.
+    pub fn try_finish(self, client: Option<Client>) -> Result<Transaction, Error> {
+        match self.clone().finish() {
+            Ok(tx) => Ok(tx),
+            Err(_) => {
+                let Some(client) = client else {
+                    return Err(Error::NoClient);
+                };
+                let Some(sender) = self.sender else {
```
rework try_finish, finish, and resolve to only have finish and resolve functions.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+}
+
+/// A transaction input that bypasses serialization -- the input contents is
+/// put verbatim into the transaction.
```
Explicitly mention that what you pass in here is already bcs serialized!

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+#[derive(Clone, Debug)]
+pub enum Input {
+    Pure(Vec<u8>),
+    // Literal(Literal),
```
```suggestion

```

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+            inputs: Vec::new(),
+            commands: Vec::new(),
+            gas: Vec::new(),
+            gas_budget: 0,
+            gas_price: 0,
```
Switch these to be `Option<u64>`

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+    /// The sponsor of the transaction. If None, the sender is also the sponsor.
+    sponsor: Option<Address>,
+    /// The expiration of the transaction.
+    expiration: Option<TransactionExpiration>,
```
get rid of Option

stefan-mysten commented:
Bring in the macros for `identifier`, `function`, `objectId`, `typetag`.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+    pub fn split_coins(&mut self, coin: Argument, amounts: Vec<Argument>) -> Argument {
+        let cmd = Command::SplitCoins(SplitCoins {
+            coin,
+            amounts: amounts.into_iter().collect(),
```
remove .into_iter()

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+    }
+
+    /// Splits a coin by the provided amounts, returning multiple results (as many as there are
+    /// amounts. To access the results, use the [`Argument::nested`] method to access the desired
```
```suggestion
    /// amounts). To access the results, use the [`Argument::nested`] method to access the desired
```

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+
+    /// Merge a list of coins into a single coin, without producing any result.
+    pub fn merge_coins(&mut self, into_coin: Argument, coins: Vec<Argument>) {
+        let cmd = Command::MergeCoins(MergeCoins {
```
fix params names to match.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+    pub fn make_move_vec(&mut self, type_: Option<TypeTag>, elements: Vec<Argument>) -> Argument {
+        let cmd = Command::MakeMoveVector(MakeMoveVector {
+            type_,
+            elements: elements.into_iter().collect(),
```
fix this

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+    /// Publish a list of modules with the given dependencies. The result is the upgrade cap, which
+    /// should be transferred to sender/another address after this call.
+    ///
+    ///  - `modules`: is the bytecode bcs encoded for the modules to be published
```
```suggestion
    ///  - `modules`: is the modules' bytecode to be published
```

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+        Argument::Result(self.commands.len() as u16 - 1)
+    }
+
+    /// Publish a list of modules with the given dependencies. The result is the upgrade cap, which
```
```suggestion
    /// Publish a list of modules with the given dependencies. The result is the `0x2::package::UpgradeCap`, which
```

stefan-mysten replied:
Reword this to mention that you need to process the upgrade cap: burn it, send it, wrap it.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1046 @@
+    ///  - `dependencies`: is the list of IDs of the transitive dependencies of the package to be
+    ///     upgraded
+    ///  - `prev`: is the ID of the current package being upgraded
+    ///  - `ticket`: is the upgrade ticket
```
Add example here how to get the upgrade ticket.

bmwill commented on crates/sui-transaction-builder/src/object.rs:
```diff
@@ -0,0 +1,205 @@
+
+/// Type representing potentially unresolved object types, with a builder API.
+#[derive(Clone, Debug)]
+pub struct Object {
```
What value does this provide that isn't already provided by `UnresolvedInputArgument`?

stefan-mysten replied:
Yeah, I don't think we need this anymore.

stefan-mysten replied:
Would it be ok to add some constructors to `UnresolvedInputArgument`. I think it's a bit nicer to do
```rust
let coin_to_merge = tx.input(Object::owned(
            coin1_obj.object_id(),
            coin1_version,
            coin1_digest,
        ));
```
rather than
```
        let a = tx.input(UnresolvedInputArgument {
            kind: Some(UnresolvedInputArgumentKind::ImmutableOrOwned),
            object_id: Some(coin1),
            version: Some(coin1_version),
            digest: Some(coin1_digest),
            mutable: None,
            value: None,
        });

```

bmwill replied:
Oh certainly, let's do that!

stefan-mysten replied:
How about `initial_shared_version` in `UnresovledInputArgument` if it is a `Shared` kind of object?

stefan-mysten replied:
I see that `version` has an alias `initial_shared_version`. That's a bit confusing to me 😅

bmwill commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1006 @@
+    /// The arguments required for this command are:
+    ///  - `modules`: is the modules' bytecode to be published
+    ///  - `dependencies`: is the list of IDs of the transitive dependencies of the package to be
+
```
extra empty line?

stefan-mysten commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -188,3 +207,57 @@ impl From<UnresolvedValue> for serde_json::Value {
+                                arr.push(b);
+                            }
+                            Ok(InputArgument::Pure { value: arr })
+                        }
+                        _ => Err("Pure expects an array but got something else instead"),
+                    }
+                }
+                UnresolvedInputArgumentKind::Shared => Ok(InputArgument::Shared {
+                    object_id: value.object_id.expect("object_id must be present"),
+                    initial_shared_version: value.version.expect("version must be present"),
+                    mutable: value.mutable.expect("mutable must be present"),
+                }),
+                UnresolvedInputArgumentKind::Receiving => {
+                    Ok(InputArgument::Receiving(ObjectReference::new(
+                        value.object_id.expect("object_id must be present"),
+                        value.version.expect("version must be present"),
+                        value.digest.expect("digest must be present"),
+                    )))
+                }
+                UnresolvedInputArgumentKind::ImmutableOrOwned
+                | UnresolvedInputArgumentKind::Owned => {
+                    Ok(InputArgument::ImmutableOrOwned(ObjectReference::new(
+                        value.object_id.expect("object_id must be present"),
+                        value.version.expect("version must be present"),
+                        value.digest.expect("digest must be present"),
+                    )))
+                }
+                UnresolvedInputArgumentKind::Immutable => {
+                    Ok(InputArgument::ImmutableOrOwned(ObjectReference::new(
+                        value.object_id.expect("object_id must be present"),
+                        value.version.expect("version must be present"),
+                        value.digest.expect("digest must be present"),
+                    )))
+                }
+                UnresolvedInputArgumentKind::Literal => todo!(),
+            },
+            None => todo!(),
+        }
+    }
+}
```
Ok, we'd need to have it as a function instead of a TryFrom because `UnresolvedInputArgument` is implemented here.

stefan-mysten replied:
Left `unresolved.rs` intact and moved this logic into the `library`.

stefan-mysten replied:
Also fixed all the improper error handling.

stefan-mysten commented on crates/sui-sdk-types/src/types/transaction/mod.rs:
```diff
@@ -626,6 +626,17 @@ pub enum Argument {
+    pub fn nested(&self, ix: u16) -> Argument {
+        let Argument::Result(i) = self else {
+            panic!("Cannot nest a non-result value");
+        };
```
Fixed!

amnn commented on crates/sui-sdk-types/src/types/transaction/mod.rs:
```diff
@@ -626,6 +626,16 @@ pub enum Argument {
 }
 
+impl Argument {
+    /// Turn a Result into a NestedResult.
```
nit: Document when this returns `None` vs `Some`.

bmwill commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1060 @@
+    // Transaction Inputs
+
+    /// Make a value available to the transaction as an input.
+    pub fn input(&mut self, i: impl Into<UnresolvedInputArgument>) -> Argument {
```
Yeah we could even rename the types themselves and expose them from the module "unresolved"

stefan-mysten replied:
Renamed all the types and using `unresolved::TypeName` wherever needed.

bmwill commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1060 @@
+    fn from(raw: RawBytes) -> Self {
+        Self {
+            kind: Some(UnresolvedInputArgumentKind::Pure),
+            value: Some(UnresolvedValue::Array(
+                raw.0
+                    .into_iter()
+                    .map(|x| UnresolvedValue::Number(x as u64))
+                    .collect(),
+            )),
```
Yeah if we know its a pure arg we should serialize it as a base64 string.

stefan-mysten replied:
Fixed.

stefan-mysten commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -150,6 +162,70 @@ pub enum UnresolvedValue {
     Array(Vec<UnresolvedValue>),
 }
 
+impl UnresolvedInputArgument {
```
Fixed

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1060 @@
+    // Metadata
+
+    /// Set the gas objects that will be used to pay for the transaction.
+    pub fn set_gas(&mut self, gas: Vec<UnresolvedObjectReference>) {
```
I think that makes sense (after reading it 5 times 😅), but still not sure about the `handle any object input` and a `type that can handle all inputs`. Isn't the `UnresovledInputArgument` that type that handles both?

amnn replied:
Yeah I think that's a good simplification (to use `UnresolvedInputArgument` for everything). On the one hand it means that people could attempt to pass a pure input here, but on the other hand, we already have to validate these inputs, and doing it this way means you could support passing literals here (in particular, you could pass an Object ID as a string and have it be understood as an ObjectID.

You may want to change the API to something like this though:

```rust
pub fn add_gas<O, I>(&mut self, gas: I)
where
    O: Into<UnresolvedInputArgument>,
    I: IntoIter<Item = O>,
{
    self.gas.extend(gas.into_iter().map(Into::into))
}
```

This is the most flexible form of the API I can think of. It allows for things like:

```rust
tx.add_gas(["0x1....2", "0x3...4"]);
tx.add_gas([prev_input]);
```

i.e. you can pass in literals without having to call `.into()` on them, to match how we handle accepting inputs via `tx.input(...)` and you can also pass multiple different kinds of gas object (inputs that are just literals, inputs that are already partially resolved objects of some kind, etc).

The alternative (with the current interface) would be that the literals would need to have explicit calls to `.into()` after them:

```rust
tx.set_gas(vec![
    "0x1....2".into(),
    "0x3....4".into(),
    prev_input
]);
```

Maybe this is fine -- will leave that part up to you.

stefan-mysten replied:
Implemented this, thanks for the suggestion.

amnn commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -93,28 +94,36 @@ pub struct UnresolvedObjectReference {
 #[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
-pub struct UnresolvedInputArgument {
+pub struct InputArgument {
+    /// The object kind.
     #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none",))]
-    pub kind: Option<UnresolvedInputArgumentKind>,
+    pub kind: Option<InputArgumentKind>,
```
nit: Can we call these `Input` and `InputKind`? I don't think `Argument` is adding much here.

stefan-mysten replied:
To quote a VIP - the world is our oyster, so yes we can :D

stefan-mysten replied:
Renamed.

stefan-mysten commented:
@bmwill @amnn thanks for all the feedback. I should have addressed all of it. The one thing that remains to do is to bring in/define some of the macros `obj!(), input!()` to make it easier to create the needed data. I will do this in a subsequent PR so that we can finally land this.

bmwill commented:
why do we need macros? I tend to have an aversion to user-facing macros as they make things more complex rather than simpler from the perspective of an end user

amnn reviewed (commented):
Couple of themes in this round of comments:

- Taking a closer look at the error story (I may be missing some details but the `error` module has now gotten quite a bit more complicated and I didn't spot a great indication as to why), e.g.
  - Why are we depending on `thiserror` but then not using it.
  - `ConversionError` is not very structured.
  - What's the motivation for the overall error type, boxing the `BuilderError`?
- Consistent use of `parse()` instead of `from_str` where possible.
- Use `TransactionEffectsAPI` to avoid having to check for the V2 effects structure all the time.

And then some larger individual comments:

- The upgrade example doesn't seem like it should work and needs another look.
- I wasn't very clear about my suggestion to use Base64 to encode pure arguments -- that only applies to when we are sending it over the wire. We shouldn't be stashing bytes in a String at rest.
- Avoiding having the raw module bytes stored inline on the stack for tests -- is there a way we could compile these as part of the test process?

Then there are some smaller one-off suggestions and comments as well.

bmwill commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+        let (address, pk, _) = helper_setup(&mut tx, &client).await;
+
+        // publish very dummy package (from sui/examples/move/first_package)
+        let modules: [u8; 506] = [
```
I'm not against having raw compiled bytes here, if we do want to compile the bytes as a part of the test then we'd need to build utilities to shell out to the move compiler or take a dep on the mono-repo for that (which would be nice to avoid)

stefan-mysten replied:
To avoid duplicating code, we could use the `sui` binary to do a `sui move build --dump-bytecode-base64` and get the compiled modules, digest, and dependencies from there. 
The advantage is that we're not copy pasting code and we're already using the CLI to start a local network for tests, so this seems reasonable to me.

stefan-mysten replied:
Let me know 👍  or 👎

amnn replied:
I'm happy with the shelling out approach, IIUC you are already using a Makefile as a test runner, so you could use that to generate test fixtures into a directory before the test starts -- seems fine to me.

amnn replied:
I think we are saying the same thing, but just to be explicit, could those instructions be a make rule that builds the Move bytecode (at the make level we need to depend on the `sui` CLI anyway to run the network), and could the test's make rule depend on that rule to make sure the fixture is always up-to-date?

That way there is no shelling out in the Rust code for the test, no dependency on the mono-repo at the crate level, but people don't have to worry about understanding how the bytes are generated and manually updating them.

stefan-mysten replied:
Let's go like this and if needed down the road, happy to add another way to handle this.

amnn replied:
It's maybe less relevant for testing publishing and upgrading but how will you test making a move call in a PTB without being able to publish a package with functions to call?

stefan-mysten replied:
Unless we need to test specific hand-rolled packages, we can use the framework to call different functions?

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+        }
+
+        // check that it succeeded
+        let status = match effects.unwrap().unwrap() {
+            sui_types::types::TransactionEffects::V2(e) => Some(e.status),
+            _ => None,
+        };
+        let expected_status = ExecutionStatus::Success;
+        assert_eq!(expected_status, status.unwrap());
```
Started building the tx effects api, so fixed this.

stefan-mysten commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -93,28 +95,35 @@ pub struct UnresolvedObjectReference {
 )]
 #[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
-pub struct UnresolvedInputArgument {
+pub struct Input {
```
Agreed, thanks for the suggestion!

bmwill commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+        let effects = client.execute_tx(vec![sig], &tx).await;
+        let mut package_id: Option<ObjectId> = None;
+        let mut created_objs = vec![];
+        if let Ok(Some(ref effects)) = effects {
```
we definitely do not want to introduce an TransactionEffectsApi trait

stefan-mysten replied:
Will do, thanks.

stefan-mysten replied:
It's a bit unclear to me what's the right way. I was thinking that this should be fine for a few fields: `status`, `epoch`, `gas_used`, `transaction_digest`, `events_digest`. For example, I added this:
```
impl TransactionEffects {
   pub fn status(&self) -> ExecutionStatus {
      match self {
       V1(e) => e.status,
       V2(e) => e.status
      }
   }
}
```

Is this what you were suggesting @bmwill or more along the lines that both V1 and V2 should have their separate accessor implementations?

bmwill replied:
Yes the top level wrapper should provide an api that is easier to use and it should directly call methods as implemented on each inner version type.

stefan-mysten replied:
Added this in PR #60

bmwill commented:
all such checks can be done at compiletime using normal const functions and would be the preferable approach

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+    ///         Identifier::from_str("authorize_upgrade").unwrap(),
+    ///         vec![],
+    ///      ),
+    ///     vec![Argument::Input(0), upgrade_policy, digest_arg],
```
You're right, it's due to copy pasting from the test. Fixed.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+
+impl<'a, T: Serialize> From<Serialized<'a, T>> for unresolved::Input {
+    fn from(value: Serialized<'a, T>) -> Self {
+        Self {
+            kind: Some(unresolved::InputKind::Pure),
+            value: Some(unresolved::Value::String(base64ct::Base64::encode_string(
+                &bcs::to_bytes(value.0).unwrap(),
+            ))),
+            object_id: None,
+            version: None,
+            digest: None,
+            mutable: None,
+        }
```
I guess this RawBytes still confuses me. 😢

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+
+/// Convert from an [`unresolved::Input`] into an [`InputArgument`] for resolving the
+/// transaction.
+fn try_from_unresolved_input_arg(value: unresolved::Input) -> Result<InputArgument, Error> {
```
Renamed. @bmwill let me know if you see any issues with this.

stefan-mysten commented on crates/sui-transaction-builder/src/error.rs:
```diff
@@ -0,0 +1,194 @@
+    }
+}
+
+impl std::fmt::Display for Error {
```
I guess it's a matter of style. I reworked this, but I think for `Error` we should keep it as we need to have our own `impl std::error::Error for Error {fn source(...)}`

bmwill replied:
I agree with @amnn here in that, we're already using thiserror for defining a few error enums so lets just continue with that to reduce the code here. I'd probably even be fine condensing the 3 different error types/enums into a single one just so that there's a single "error type" exported from this crate

stefan-mysten replied:
I reworked a bit the errors but I am not 100% satisfied with it.

stefan-mysten replied:
@bmwill do we want to keep the pattern of having a general `Error` type that wraps an Error in a Box?

stefan-mysten commented:
> all such checks can be done at compiletime using normal const functions and would be the preferable approach

Right, cool. Would that mean that we need to reimplement `Address::from_hex` as as const function?

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+                    inputs: self
+                        .inputs
+                        .into_iter()
+                        .map(try_from_unresolved_input_arg)
```
`unresolved::Input` is defined in another crate, so we could do it there, but @bmwill suggested that we keep that as "clean" as possible, so decided to use a free function here where we need it.

bmwill replied:
we could probably look into adding that impl. But this is fine for now :)

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+                    objects: self
+                        .gas
+                        .into_iter()
+                        .map(try_from_gas_unresolved_input_to_unresolved_obj_ref)
```
Totally, will have to look at how REST API handles this stuff.

bmwill replied:
yeah the resolve api was a first pass we can and should look at iterating if we think it'll improve or help things.

For this particular one, I'm not sure if it makes sense to change what's accepted (willing to be convinced though). Given this is a gas payment we know up front that it has to be an object reference so we should on the client side be able to take a literal and make it an object id.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1091 @@
+    /// Attempts to resolve the transaction with the data provided so far by querying the GraphQL
+    /// service.
+    // TODO: finish the implementation when REST API is ready to provide the transaction resolution
+    pub fn resolve(&self, _client: &Client) -> Result<(), Error> {
```
Yes, but now we don't have the glue code so we either comment this function out or return ().

amnn replied:
I think it's fine to comment the function out (or not include it at all) until it can occupy its expected signature.

stefan-mysten commented on crates/sui-sdk-types/src/types/effects/v1.rs:
```diff
@@ -15,7 +15,7 @@ use crate::types::TransactionEventsDigest;
     /// The status of the execution
     #[cfg_attr(feature = "schemars", schemars(flatten))]
-    status: ExecutionStatus,
+    pub status: ExecutionStatus,
```
All fields in V2 are pub while none are pub in V1. We should probably open them up?

stefan-mysten replied:
Reworked this in #60

stefan-mysten commented:
> Couple of themes in this round of comments:
> 
>     * Taking a closer look at the error story (I may be missing some details but the `error` module has now gotten quite a bit more complicated and I didn't spot a great indication as to why), e.g.

The error story changed a bit to allow more flexible errors on the user side. We're experimenting a bit on what's the right way to propagate errors, and how should the error story look like.

bmwill commented:
While const functions would improve things, i think that it should be out of scope for this particular pr

amnn commented:
Agreed, I think the only open question for me in this PR are some of the decisions around the error story. Accessors for TransactionEffects and the helpers for generating IDs/addresses/types/functions should wait for follow-ups.

jymchng commented:
Not sure why @bmwill thinks 'macros' add complexity for the end users; On the contrary, I believe macros simplify things for the end users: `object_id!(..)` simply takes a string literal and expands it to all the const functions, sure, it does nothing more than that but it does save the end users some time from searching through the SDK which are the const functions that do the job. I mean without the macro, he might not even realize that `ObjectID` can be gotten as a const value.

bmwill commented on crates/sui-sdk-types/address.rs:
```diff
@@ -0,0 +1,263 @@
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
```
this file seems like it was added by mistake?

Given how long this PR is, it may be beneficial to squash all the commits yourself to make it a bit easier to work with

stefan-mysten replied:
Very likely, I will do a check given that I rebased on #60, maybe I missed something there.

stefan-mysten commented:
@amnn @bmwil I added a basic Move project and changed the CI and Makefile to build it and get the modules, dependencies, and digest from the build instead of hard coding them.
It's unfortunate that `sui move build --dump` requires a `client.yaml` file...

I think I've addressed most comments and the only thing that's left is errors.

bmwill commented:
Have you rebased on the latest tip of master?

bmwill commented on scripts/cli/client.yaml:
```diff
@@ -0,0 +1,15 @@
+---
+keystore:
+  File: home/.sui/sui_config/sui.keystore
+envs:
+  - alias: localnet
+    rpc: "http://127.0.0.1:9000"
+    ws: ~
+  - alias: devnet
+    rpc: "https://fullnode.devnet.sui.io:443"
+    ws: ~
+  - alias: testnet
+    rpc: "https://fullnode.testnet.sui.io:443"
+    ws: ~
+active_env: testnet
+active_address: "0x14e7ac25259adcc373c96627893976d4fe562a3f3fedce493fc187c5ebd53eee"
```
Is there a way we could not include this? Seems odd to have an active address for something that we do not keys for locally

stefan-mysten replied:
Ooouf, I hate it too. Running `sui move build --dump-bytecode-as-base64` requires this `client` file. Per the comment there
```
`sui move build` does not ordinarily require a network connection.
The exception is when --dump-bytecode-as-base64 is specified: In this
case, we should resolve the correct addresses for the respective chain
(e.g., testnet, mainnet) from the Move.lock under automated address management.
```

stefan-mysten replied:
We could change this to allow passing in an `unknown` chain parameter to avoid having the `client.yaml` file. The issue is that then we'd have to wait for the testnet release as that's the binary that we're downloading for running tests.

bmwill replied:
or is there a way to just init it without needing to have a file to copy?

stefan-mysten replied:
Added a flag in the `sui move build` command to ignore this. https://github.com/MystenLabs/sui/pull/20475

stefan-mysten replied:
Generated on the fly and added a issue to later switch to `--ignore-chain` and avoid this thing altogether.

bmwill commented on crates/sui-transaction-builder/Makefile:
```diff
@@ -0,0 +1,3 @@
+test-setup:
+	cd tests/test_example_v1 && sui move build --dump-bytecode-as-base64 >> ../../test_example_v1_build_output.json
+	cd tests/test_example_v2 && sui move build --dump-bytecode-as-base64 >> ../../test_example_v2_build_output.json 
```
I don't see these files either:
- added to the repo (which if we're generating them we don't need)
- ignored via .gitignore so that they don't get added to the repo.

stefan-mysten replied:
When you refer to these files you're suggesting these two files `test_example_v1_build_output.json` + v2?
We don't need to check them in, they are generated on each run? 

I will add them to `.gitignore`.

amnn reviewed (commented):
A small issue with `IntoInput` and in particular the way it's implemented for `Object`, but otherwise this is good to go in my book! 🚀 well done @stefan-mysten

bmwill commented on Makefile:
```diff
@@ -26,7 +26,8 @@ test:
 
 .PHONY: test-with-localnet
 test-with-localnet:
-	cargo nextest run -p sui-graphql-client
+	$(MAKE) -C crates/sui-transaction-builder test-setup
+	cargo nextest run -p sui-graphql-client -p sui-transaction-builder
```
only change would be that I think `.PHONY` would need to be removed from `test-with-localnet` now since it actually does have a dependency based on files on disk

amnn replied:
To me `.PHONY` means "always run this rule, regardless of whether its rules are up-to-date or not" (which we want because we do always want to re-run the tests), but specifying the dependencies will ensure that if the JSON files are out-of-date w.r.t. to their respective inputs, they will be regenerated.

Having said that though, I now realise that we need to properly specify the dependencies in the `package_%.json` rule -- it needs to depend on everything inside the relevant test fixture folder.

stefan-mysten replied:
I think I fixed this as well.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1008 @@
+/// A transaction builder to build transactions.
+impl TransactionBuilder {
+    /// Create a new transaction builder and initialize its elements to default.
+    /// This sets the transaction expiration to have no expiration. Use `set_expiration` to set the
+    /// transaction expiration to a specific epoch.
```
No logical reasoning why I did that 🤷  Let's keep it to default and users can look it up.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1008 @@
+    }
+
+    // Commands
+    /// Call a public or a Move function with the given arguments.
```
Oh, that's useful to know!

stefan-mysten commented on crates/sui-transaction-builder/src/error.rs:
Agreed, removed the doc comments.

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1008 @@
+impl IntoInput for Object {
+    /// Convert the [`Object`] type into an [`unresolved::Input`] of kind immutable.
+    fn into_input(self) -> unresolved::Input {
+        unresolved::Input::immutable(self.object_id(), self.version(), self.digest())
```
Thanks for catching this. I somehow missed it. Fixed everything!

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1008 @@
+        println!("Coins: {:?}", coins);
+        let gas = coins.last().unwrap().id;
+        let gas_obj = client.object(gas.into(), None).await.unwrap().unwrap();
+        tx.add_gas(vec![gas_obj]);
```
In theory yes that should work. In practice, now, it will fail because it requires to figure out the version and digest of an object and we don't have the transaction resolution plugged in yet.

stefan-mysten replied:
After fixing `IntoInput` impl, passing `&gas_obj` will not work either because the object has an `Address` owner and we do not know if it's owned or receiving type, so we have to build the object by passing the version and digest in addition to the object id.

amnn replied:
In this case, maybe it's worth having a special function on `Object` to convert it into an `Input` that you can then modify (see the discussion in my latest round of review comments on the `unresolved::Input` builder APIs):

```rust
impl Object {
    pub fn as_input(&self) -> unresolved::Input;
}

tx.add_gas([gas_obj.as_input().with_owned_kind()]);
```

stefan-mysten replied:
We'd still need to get the object because we need version + digest. Object itself does not have a digest, so we'd have to compute it / add a function on Object to compute it.

stefan-mysten replied:
Sorry, Object does have a digest that can be computed already. Hmm.

stefan-mysten replied:
Added it!

amnn reviewed (commented):
Final round of comments from me, I think -- the latest round of changes opened a Pandora's box around `unresolved::Input` (`IntoInput` + the builder API mainly).

stefan-mysten commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,1008 @@
+/// A trait to convert a type into an [`unresolved::Input`]. This is mostly used for gas objects,
+/// to pass them either as literals, as an object id, as a reference to an object, or as a
+/// [`unresolved::Input`] object.
+pub trait IntoInput {
```
Maybe I am doing something wrong or maybe it's a new rule in Rust 1.81?

```rust
error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
   --> crates/sui-transaction-builder/src/lib.rs:402:1
    |
402 | impl Into<unresolved::Input> for &Object {
    | ^^^^^-----------------------^^^^^-------
    | |    |                           |
    | |    |                           `sui_sdk_types::types::Object` is not defined in the current crate
    | |    `sui_sdk_types::types::unresolved::Input` is not defined in the current crate
    | impl doesn't use only types from inside the current crate
    |
    = note: define and implement a trait or new type instead
```

stefan-mysten replied:
Move this into `unresolved` module.

amnn replied:
(We spoke about this offline, but for posterity the issue here is that `Object` is from a different crate, not `Input`)

stefan-mysten commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -207,12 +207,12 @@ impl Input {
     /// - `mutable` controls whether a command can accept the object by value or mutable reference.
     /// - `initial_shared_version` is the first version the object was shared at.
-    pub fn shared(object_id: ObjectId, initial_shared_version: u64, mutable: bool) -> Self {
+    pub fn shared(object_id: ObjectId, initial_shared_version: u64, mutable: Option<bool>) -> Self {
```
Ah yes, don't know why I changed it.

stefan-mysten replied:
Fixed!

stefan-mysten commented on Makefile:
```diff
@@ -24,9 +24,12 @@ test:
 	cargo nextest run --all-features -p sui-sdk-types -p sui-crypto
 	cargo test --doc
 
+package_%.json:
```
Tried to add something but not sure how to test it. Also, because test-with-localnet is phony, I think this rule will always be called?

amnn commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -226,60 +228,91 @@ impl Input {
+    pub fn with_initial_shared_version(self, initial_version: u64) -> Self {
+        Self {
+            version: Some(initial_version),
+            kind: Some(InputKind::Shared),
```
minor nit: order (`kind` is first in all other cases).

amnn commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -321,3 +354,27 @@ impl From<Value> for serde_json::Value {
 }
+
+impl From<&Object> for Input {
+    fn from(object: &Object) -> Self {
```
This should call `object.as_input()` and `as_input`'s body should contain this logic.

amnn commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -321,3 +354,27 @@ impl From<Value> for serde_json::Value {
+    }
+}
+
+impl From<&ObjectId> for Input {
```
Does `ObjectId` implement `Copy`? If so, we should take it by value here.

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -964,7 +919,7 @@ mod tests {
                         }
                         sui_types::types::Owner::Shared(x) => {
-                            upgrade_cap = Some(tx.input(unresolved::Input::shared(o, *x, Some(true))));
+                            upgrade_cap = Some(tx.input(unresolved::Input::shared(o, *x, true)));
```
Can you use `tx.input(&o)` here too now?

stefan-mysten replied:
Yesss 🕺

amnn commented on crates/sui-transaction-builder/src/lib.rs:
```diff
@@ -0,0 +1,981 @@
+    type_args: Vec<TypeTag>,
+}
+
+/// A trait to convert a type into an [`unresolved::Input`]. This is mostly used for gas objects,
+/// to pass them either as literals, as an object id, as a reference to an object, or as a
+/// [`unresolved::Input`] object.
+// pub trait IntoInput {
+//     /// Convert the type into an [`unresolved::Input`].
+//     fn into_input(self) -> unresolved::Input;
+// }
```
Flagging for deletion

amnn reviewed (commented):
Okay that's everything!! 🚀 From my perspective, we're good to go. Thanks for sticking with this @stefan-mysten, I think people are going to really enjoy using this new Transaction Builder.

Will leave @bmwill to give the final stamp

stefan-mysten commented:
> Okay that's everything!! 🚀 From my perspective, we're good to go. Thanks for sticking with this @stefan-mysten, I think people are going to really enjoy using this new Transaction Builder.
> 
> Will leave @bmwill to give the final stamp

Thanks so much for all the in-depth reviews and for your endless patience @amnn!

bmwill commented on crates/sui-sdk-types/src/types/object.rs:
```diff
@@ -1,5 +1,7 @@
 use std::collections::BTreeMap;
 
+#[cfg(all(feature = "serde", feature = "hash"))]
+use super::unresolved::Input;
```
Give this is used in only a single place, lets inline the import to avoid needing to handle the conditional compilation in two places.

bmwill commented on crates/sui-sdk-types/src/types/object.rs:
```diff
@@ -322,6 +324,14 @@ impl Object {
+
+    /// Convert this object into an [`unresolved::Input`] with the object id, digest, and version.
+    #[cfg(all(feature = "serde", feature = "hash"))]
+    pub fn as_input(&self) -> Input {
+        Input::by_id(self.object_id())
```
```suggestion
    pub fn as_input(&self) -> super::unresolved::Input {
        super::unresolved::Input::by_id(self.object_id())
```

bmwill replied:
I'm not actually a huge fan of this approach. honestly this should be handled by the TxnBuilder and not done here. I know there was some reasons with implementing `From` but maybe we end up moving the unresolved types out at some point.

This is just a comment on my preferences for having a cleaner api and something we can iterate on later (doesn't need to be fix now)

bmwill commented on crates/sui-sdk-types/src/types/transaction/unresolved.rs:
```diff
@@ -321,3 +356,21 @@ impl From<Value> for serde_json::Value {
     }
 }
+
+#[cfg(all(feature = "serde", feature = "hash"))]
+impl From<&Object> for Input {
+    fn from(object: &Object) -> Self {
+        match object.owner() {
+            Owner::Address(_) => object.as_input(),
+            Owner::Object(_) => object.as_input().with_owned_kind(),
+            Owner::Shared(at_version) => object.as_input().with_initial_shared_version(*at_version),
+            Owner::Immutable => object.as_input().with_immutable_kind(),
+        }
+    }
+}
```
I'm having a really hard time understanding why we'd have a From impl here that doesn't match the as_input impl above. This seems very confusing and error prone. Lets get rid of one of the two of these (maybe the as_input method).

Same comment as above, inline the imports here either by fully qualifying the paths or by just having the use statements inside the function.
